src/main/java/net/fabricmc/loader/gui/FabricStatusTree.java

```diff
@@ -0,0 +1,310 @@
+			this.shouldContinue = true;
+			return this;
+		}
+	}
+
+	public static final class FabricStatusTab {
+		public final FabricStatusNode node;
+
+		/** The minimum warning level to display for this tab. */
+		public FabricTreeWarningLevel filterLevel = FabricTreeWarningLevel.NONE;
+
+		public FabricStatusTab(String name) {
+			this.node = new FabricStatusNode(null, name);
+		}
+
+		public FabricStatusNode addChild(String name) {
+			return node.addChild(name);
+		}
+	}
+
+	public static final class FabricStatusNode {
+
+		private transient FabricStatusNode parent;
+
+		public String name;
+
+		/** The icon type. There can be a maximum of 2 decorations (added with "+" symbols), or 3 if the
+		 * {@link #setWarningLevel(FabricTreeWarningLevel) warning level} is set to
+		 * {@link FabricTreeWarningLevel#NONE } */
+		public String iconType = ICON_TYPE_DEFAULT;
+
+		private FabricTreeWarningLevel warningLevel = FabricTreeWarningLevel.NONE;
+
+		public boolean expandByDefault = false;
+
+		public final List<FabricStatusNode> children = new ArrayList<>();
+
+		/** Extra text for more information. Lines should be separated by "\n". */
+		public String details;
+
+		private FabricStatusNode(FabricStatusNode parent, String name) {
+			this.parent = parent;
+			this.name = name;
+		}
+
+		public void moveTo(FabricStatusNode newParent) {
+			parent.children.remove(this);
+			this.parent = newParent;
+			newParent.children.add(this);
+		}
+
+		public FabricTreeWarningLevel getMaximumWarningLevel() {
+			return warningLevel;
+		}
+
+		public void setWarningLevel(FabricTreeWarningLevel level) {
+			if (this.warningLevel == level) {
+				return;
+			}
+
+			if (warningLevel.isWorseThan(level)) {
+				// Just because I haven't written the back-fill revalidation for this
+				throw new Error("Why would you set the warning level multiple times?");
+			} else {
+				if (parent != null && level.isWorseThan(parent.warningLevel)) {
+					parent.setWarningLevel(level);
+				}
+
+				this.warningLevel = level;
+			}
+		}
+
+		public void setError() {
+			setWarningLevel(FabricTreeWarningLevel.ERROR);
+		}
+
+		public void setWarning() {
+			setWarningLevel(FabricTreeWarningLevel.WARN);
+		}
+
+		public void setInfo() {
+			setWarningLevel(FabricTreeWarningLevel.INFO);
+		}
+
+		public FabricStatusNode addChild(String string) {
+			FabricStatusNode child = new FabricStatusNode(this, string);
+			children.add(child);
+			return child;
+		}
+
+		/** Calls {@link #addException(Throwable)}, and throws the given exception */
+		public <T extends Throwable> T addAndThrow(T exception) throws T {
+			addException(exception);
+			throw exception;
+		}
+
+		public FabricStatusNode addException(Throwable exception) {
+			FabricStatusNode sub = new FabricStatusNode(this, "...");
+			children.add(sub);
+
+			sub.setError();
+			String msg = exception.getMessage();
+			String[] lines = msg.split("\n");
+
+			if (lines.length == 0) {
+				// what
```

Comment:
If the message only contains newlines then `split("\n")` seems to return an empty array. Otherwise it seems to always return at least 1 element (even for the empty string it returns an array containing a single empty string).

But yes this could be improved.